### PR TITLE
EBL-162- Block by availability

### PR DESCRIPTION
**Pull request by mayhuberman**

### CODE REVIEW

File: apollo-backend/src/main/java/io/logz/apollo/controllers/BlockerDefinitionController.java [hunk 1]
```diff
@@ -172,6 +148,24 @@ public void updateBlockerDefinition(int id, String name, String environmentId, S
         assignJsonResponseToReq(req, HttpStatus.OK, blockerDefinition);
     }
 
+    private boolean isValid(String availability, Integer stackId, Integer environmentId, Integer serviceId) {
+        if (stackId != null && !stackId.equals("null")) {
+            if (environmentId != null || serviceId != null) {
```
Comment (roiravhon) [inline apollo-backend/src/main/java/io/logz/apollo/controllers/BlockerDefinitionController.java]: I can have environment stack and a service ID, or service stack and environment ID

File: apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java [hunk 2]
```diff
@@ -198,6 +198,80 @@ public void testServicesStackBlocker() throws Exception {
...
+        Service service = ModelsGenerator.createAndSubmitService(apolloTestClient);
+
+        createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", null, null, null, availability);
+        DeployableVersion deployableVersion = ModelsGenerator.createAndSubmitDeployableVersion(apolloTestClient, service);
+        assertThatThrownBy(() -> ModelsGenerator.createAndSubmitDeployment(apolloTestClient, environment, service, deployableVersion)).isInstanceOf(Exception.class)
+                                                                                                                                         .hasMessageContaining("Deployment is currently blocked");
+    }
+
+    @Test
+    public void testNonBlockedDeploymentWithAvailabilityBlockerOnSpecificService() throws Exception {
```
Comment (roiravhon) [inline apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java]: Remove this, and add another line  before each assert that check for the "pass" version

File: apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java [hunk 3]
```diff
@@ -198,6 +198,80 @@ public void testServicesStackBlocker() throws Exception {
...
+        Service blockedService = ModelsGenerator.createAndSubmitService(apolloTestClient);
+        Service service = ModelsGenerator.createAndSubmitService(apolloTestClient);
+
+        createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", null, blockedService, null, availability);
+        DeployableVersion deployableVersion = ModelsGenerator.createAndSubmitDeployableVersion(apolloTestClient, service);
+        ModelsGenerator.createAndSubmitDeployment(apolloTestClient, environment, service, deployableVersion);
+    }
+
+    @Test
+    public void testNonBlockedDeploymentWithAnotherAvailabilityBlocker() throws Exception {
```
Comment (roiravhon) [inline apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java]: Same

File: apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java [hunk 4]
```diff
@@ -198,6 +198,80 @@ public void testServicesStackBlocker() throws Exception {
...
+        ApolloTestClient apolloTestClient = Common.signupAndLogin();
+        ApolloTestAdminClient apolloTestAdminClient = Common.getAndLoginApolloTestAdminClient();
+
+        String availability = "BLOCK-ME-5";
+        Environment environment = ModelsGenerator.createAndSubmitEnvironment(availability, apolloTestClient);
+        String errorMessage = String.format("Trying to add invalid blocker. stackId - null, environmentId - %s, serviceId - null, availability - %s", environment.getId(), availability);
+
+        assertThatThrownBy(() -> ModelsGenerator. createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", environment, null, null, availability))
+                .isInstanceOf(ApolloClientException.class)
+                .hasMessageContaining(errorMessage);
```
Comment (roiravhon) [inline apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java]: .hasMessageContaining("Trying to add invalid blocker");

File: apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java [hunk 5]
```diff
@@ -207,10 +281,10 @@ public void testAddingInvalidBlockers() throws Exception {
...
         Stack servicesStack = ModelsGenerator.createAndSubmitServicesStack(apolloTestClient, Arrays.asList(service.getId()));
 
-        String errorMessage = "Trying to add a stack blocker that is also an environment or a service blocker. ";
+        String errorMessage = "Trying to add invalid blocker. ";
 
         assertThatThrownBy(() -> ModelsGenerator. createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", null, service, servicesStack)).isInstanceOf(ApolloClientException.class)
-                                                                                                                                                                   .hasMessageContaining(errorMessage + String.format("stackId - %s, environmentId - null, serviceId - %s", servicesStack.getId(), service.getId()));
+                                                                                                                                                                   .hasMessageContaining(errorMessage + String.format("stackId - %s, environmentId - null, serviceId - %s, availability - null", servicesStack.getId(), service.getId()));
```
Comment (roiravhon) [inline apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java]: .hasMessageContaining(errorMessage)

File: apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java [hunk 6]
```diff
@@ -207,10 +281,10 @@ public void testAddingInvalidBlockers() throws Exception {
...
 
         assertThatThrownBy(() -> ModelsGenerator. createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", environment, null, servicesStack)).isInstanceOf(ApolloClientException.class)
```
Comment (roiravhon) [inline apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java]: It should pass

File: apollo-client/src/main/java/io/logz/apollo/clients/ApolloAdminClient.java [hunk 7]
```diff
@@ -148,4 +150,12 @@ private String generateSignupJson(User user, String plainPassword) {
                                     "userEmail", user.getUserEmail(),
                                     "password", plainPassword);
     }
+
+    private String transferToString(Object object) {
```
Comment (roiravhon) [inline apollo-client/src/main/java/io/logz/apollo/clients/ApolloAdminClient.java]: name change?

File: apollo-backend/src/main/java/io/logz/apollo/controllers/BlockerDefinitionController.java [hunk 8]
```diff
@@ -172,6 +178,18 @@ public void updateBlockerDefinition(int id, String name, String environmentId, S
...
+    private void checkAvailabilityBlockerValidation(String availability, String environmentId, String methodType, Req req) {
+        if (availability == null || availability.isEmpty()) {
+            return;
+        } else {
+            if (environmentId == null || environmentId.equals("null")) {
+                return;
+            }
+        }
+        assignJsonResponseToReq(req, HttpStatus.BAD_REQUEST, String.format("Trying to %s an availability blocker that is also an environment blocker. availability - %s, environmentId - %s", methodType, availability, environmentId));
+        throw new RuntimeException(String.format("Trying to %s an availability blocker that is also an environment blocker. availability - %s, environmentId - %s", methodType, availability, environmentId));
```
[the remark this answers is not in the document]
Comment (mayhuberman) [reply apollo-backend/src/main/java/io/logz/apollo/controllers/BlockerDefinitionController.java]: Removed already

File: apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java [hunk 9]
```diff
@@ -198,6 +198,68 @@ public void testServicesStackBlocker() throws Exception {
...
+
+        String availability = "BLOCK-ME-1";
+        Environment environment1 = ModelsGenerator.createAndSubmitEnvironment(availability, apolloTestClient);
+        Service blockedService = ModelsGenerator.createAndSubmitService(apolloTestClient);
+
+        createAndSubmitBlocker(apolloTestAdminClient, "unconditional", "{}", null, blockedService, null, availability);
+        DeployableVersion deployableVersion1 = ModelsGenerator.createAndSubmitDeployableVersion(apolloTestClient, blockedService);
+        assertThatThrownBy(() -> ModelsGenerator.createAndSubmitDeployment(apolloTestClient, environment1, blockedService, deployableVersion1)).isInstanceOf(Exception.class)
+                                                                                                                                                           .hasMessageContaining("Deployment is currently blocked");
+        availability = "BLOCK-ME-2";
```
[the remark this answers is not in the document]
Comment (mayhuberman) [reply apollo-backend/src/test/java/io/logz/apollo/BlockerTest.java]: Splitted already

**Pull request merged**